File: services/team-service/internal/domain/events.go
package domain

const (
	SubjectTimeDayAdvanced  = "tiempo.dia_avanzado"
	SubjectMatchScheduled   = "partido.programado"
	SubjectMatchStarting    = "partido.iniciando"
	SubjectMatchFinished    = "partido.terminado"
	SubjectRosterPatchDelta = "roster.patch"
	SubjectSeasonPatchDelta = "season.patch"
	SubjectMatchResultDelta = "match.result"
)

type EventMeta struct {
	EventID       string `json:"event_id"`
	GameID        string `json:"game_id"`
	OccurredAt    string `json:"occurred_at"`
	SchemaVersion uint16 `json:"schema_version"`
}

type DayAdvancedEvent struct {
	EventMeta
	SimulatedDate string `json:"simulated_date"`
	Speed         uint8  `json:"speed"`
	DaysProcessed uint16 `json:"days_processed"`
}

type MatchScheduledEvent struct {
	EventMeta
	MatchID       string               `json:"match_id"`
	SimulatedDate string               `json:"simulated_date"`
	HomeTeam      MatchTeam            `json:"home_team"`
	AwayTeam      MatchTeam            `json:"away_team"`
	HomeTactics   MatchTacticalContext `json:"home_tactics"`
	AwayTactics   MatchTacticalContext `json:"away_tactics"`
	Players       []MatchPlayer        `json:"players"`
	Seed          uint64               `json:"seed"`
}

type MatchFinishedEvent struct {
	EventMeta
	MatchID       string           `json:"match_id"`
	SimulatedDate string           `json:"simulated_date"`
	HomeTeam      MatchTeam        `json:"home_team"`
	AwayTeam      MatchTeam        `json:"away_team"`
	HomeScore     uint16           `json:"home_score"`
	AwayScore     uint16           `json:"away_score"`
	WinnerTeamID  string           `json:"winner_team_id"`
	Seed          uint64           `json:"seed"`
	BoxScore      []PlayerBoxScore `json:"box_score"`
	KeyMoments    []KeyMoment      `json:"key_moments"`
}

type RosterPatchEnvelope struct {
	Type    string           `json:"type"`
	Subject string           `json:"subject"`
	GameID  string           `json:"game_id"`
	Patch   RosterStatePatch `json:"patch"`
}

type RosterStatePatch struct {
	SimulatedDate string                 `json:"simulated_date"`
	SourceEventID string                 `json:"source_event_id"`
	SourceSubject string                 `json:"source_subject"`
	Players       []PlayerEmotionalPatch `json:"players"`
}

type PlayerEmotionalPatch struct {
	PlayerID         string  `json:"player_id"`
	EmotionalState   string  `json:"emotional_state"`
	Satisfaction     float64 `json:"satisfaction"`
	Loyalty          float64 `json:"loyalty"`
	Ego              float64 `json:"ego"`
	CompetitiveDrive float64 `json:"competitive_drive"`
	CityConnection   float64 `json:"city_connection"`
	Summary          string  `json:"summary"`
}

type SeasonPatchEvent struct {
	Type    string           `json:"type"`
	Subject string           `json:"subject"`
	GameID  string           `json:"game_id"`
	Patch   SeasonStatePatch `json:"patch"`
}

type SeasonStatePatch struct {
	Wins          uint16              `json:"wins"`
	Losses        uint16              `json:"losses"`
	PointsFor     uint16              `json:"points_for"`
	PointsAgainst uint16              `json:"points_against"`
	LastResult    *SeasonMatchSummary `json:"last_result,omitempty"`
}

type SeasonMatchSummary struct {
	MatchID       string `json:"match_id"`
	SimulatedDate string `json:"simulated_date"`
	HomeTeamID    string `json:"home_team_id"`
	AwayTeamID    string `json:"away_team_id"`
	HomeScore     uint16 `json:"home_score"`
	AwayScore     uint16 `json:"away_score"`
	WinnerTeamID  string `json:"winner_team_id"`
}

type SeasonRecord struct {
	GameID        string
	Wins          uint16
	Losses        uint16
	PointsFor     uint16
	PointsAgainst uint16
	LastResult    *SeasonMatchSummary
}

func SeasonPatchFromRecord(record SeasonRecord) SeasonPatchEvent {
	return SeasonPatchEvent{
		Type:    SubjectSeasonPatchDelta,
		Subject: SubjectSeasonPatchDelta,
		GameID:  record.GameID,
		Patch: SeasonStatePatch{
			Wins:          record.Wins,
			Losses:        record.Losses,
			PointsFor:     record.PointsFor,
			PointsAgainst: record.PointsAgainst,
			LastResult:    record.LastResult,
		},
	}
}

type MatchTeam struct {
	TeamID             string `json:"team_id"`
	Name               string `json:"name"`
	Abbreviation       string `json:"abbreviation"`
	Rating             uint8  `json:"rating"`
	OffenseRating      uint8  `json:"offense_rating"`
	DefenseRating      uint8  `json:"defense_rating"`
	Pace               uint8  `json:"pace"`
	HomeCourtAdvantage int8   `json:"home_court_advantage"`
}

type MatchTacticalContext struct {
	System             string `json:"system"`
	RotationPreference string `json:"rotation_preference"`
	Flexibility        uint8  `json:"flexibility"`
}

type MatchPlayer struct {
	PlayerID        string `json:"player_id"`
	TeamID          string `json:"team_id"`
	ExpectedMinutes uint8  `json:"expected_minutes"`
	Rating          uint8  `json:"rating"`
	Scoring         uint8  `json:"scoring"`
	Rebounding      uint8  `json:"rebounding"`
	Playmaking      uint8  `json:"playmaking"`
	Defense         uint8  `json:"defense"`
	Stamina         uint8  `json:"stamina"`
	Fatigue         uint8  `json:"fatigue"`
	EmotionalState  int8   `json:"emotional_state"`
}

type PlayerBoxScore struct {
	PlayerID  string `json:"player_id"`
	TeamID    string `json:"team_id"`
	Minutes   uint8  `json:"minutes"`
	Points    uint16 `json:"points"`
	Rebounds  uint16 `json:"rebounds"`
	Assists   uint16 `json:"assists"`
	Steals    uint16 `json:"steals"`
	Blocks    uint16 `json:"blocks"`
	Turnovers uint16 `json:"turnovers"`
}

type KeyMoment struct {
	Quarter     uint8  `json:"quarter"`
	Clock       string `json:"clock"`
	Kind        string `json:"kind"`
	Description string `json:"description"`
	TeamID      string `json:"team_id"`
	PlayerID    string `json:"player_id,omitempty"`
}

File: services/match-service/src/events.rs
use serde::{Deserialize, Serialize};

pub const SUBJECT_MATCH_SCHEDULED: &str = "partido.programado";
pub const SUBJECT_MATCH_STARTING: &str = "partido.iniciando";
pub const SUBJECT_MATCH_FINISHED: &str = "partido.terminado";

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct EventMeta {
    pub event_id: String,
    pub game_id: String,
    pub occurred_at: String,
    pub schema_version: u16,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MatchScheduledEvent {
    #[serde(flatten)]
    pub meta: EventMeta,
    pub match_id: String,
    pub simulated_date: String,
    pub home_team: MatchTeam,
    pub away_team: MatchTeam,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub home_tactics: Option<MatchTacticalContext>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub away_tactics: Option<MatchTacticalContext>,
    pub players: Vec<MatchPlayer>,
    pub seed: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MatchStartingEvent {
    #[serde(flatten)]
    pub meta: EventMeta,
    pub match_id: String,
    pub simulated_date: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MatchFinishedEvent {
    #[serde(flatten)]
    pub meta: EventMeta,
    pub match_id: String,
    pub simulated_date: String,
    pub home_team: MatchTeam,
    pub away_team: MatchTeam,
    pub home_score: u16,
    pub away_score: u16,
    pub winner_team_id: String,
    pub seed: u64,
    pub box_score: Vec<PlayerBoxScore>,
    pub key_moments: Vec<KeyMoment>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MatchTeam {
    pub team_id: String,
    pub name: String,
    pub abbreviation: String,
    pub rating: u8,
    pub offense_rating: u8,
    pub defense_rating: u8,
    pub pace: u8,
    pub home_court_advantage: i8,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MatchTacticalContext {
    pub system: String,
    pub rotation_preference: String,
    pub flexibility: u8,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MatchPlayer {
    pub player_id: String,
    pub team_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_minutes: Option<u8>,
    pub rating: u8,
    pub scoring: u8,
    pub rebounding: u8,
    pub playmaking: u8,
    pub defense: u8,
    pub stamina: u8,
    pub fatigue: u8,
    pub emotional_state: i8,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PlayerBoxScore {
    pub player_id: String,
    pub team_id: String,
    pub minutes: u8,
    pub points: u16,
    pub rebounds: u16,
    pub assists: u16,
    pub steals: u16,
    pub blocks: u16,
    pub turnovers: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct KeyMoment {
    pub quarter: u8,
    pub clock: String,
    pub kind: String,
    pub description: String,
    pub team_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub player_id: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn subjects_follow_m2_contracts() {
        assert_eq!(SUBJECT_MATCH_SCHEDULED, "partido.programado");
        assert_eq!(SUBJECT_MATCH_STARTING, "partido.iniciando");
        assert_eq!(SUBJECT_MATCH_FINISHED, "partido.terminado");
    }

    #[test]
    fn match_scheduled_event_accepts_m2_payload_without_tactics() {
        let payload = r#"{
            "event_id": "event-1",
            "game_id": "game-1",
            "occurred_at": "2026-10-22T00:00:00Z",
            "schema_version": 1,
            "match_id": "match-1",
            "simulated_date": "2026-10-22",
            "home_team": {
                "team_id": "home",
                "name": "Home",
                "abbreviation": "HOM",
                "rating": 78,
                "offense_rating": 79,
                "defense_rating": 76,
                "pace": 99,
                "home_court_advantage": 3
            },
            "away_team": {
                "team_id": "away",
                "name": "Away",
                "abbreviation": "AWY",
                "rating": 77,
                "offense_rating": 76,
                "defense_rating": 78,
                "pace": 97,
                "home_court_advantage": 2
            },
            "players": [],
            "seed": 42
        }"#;

        let event: MatchScheduledEvent = serde_json::from_str(payload).expect("valid m2 payload");

        assert_eq!(event.home_tactics, None);
        assert_eq!(event.away_tactics, None);
    }
}
